sdk/go/internal/types.go

```diff
@@ -225,21 +225,20 @@ func (o *OutputState) fulfillValue(value reflect.Value, known, secret bool, deps
 }
 
 func mergeDependencies(ours []Resource, theirs []Resource) []Resource {
-	if len(ours) == 0 && len(theirs) == 0 {
+	total := len(ours) + len(theirs)
+	if total == 0 {
 		return nil
-	} else if len(theirs) == 0 {
-		return append(slice.Prealloc[Resource](len(ours)), ours...)
-	} else if len(ours) == 0 {
-		return append(slice.Prealloc[Resource](len(theirs)), theirs...)
 	}
-	depSet := make(map[Resource]struct{})
-	mergedDeps := slice.Prealloc[Resource](len(ours) + len(theirs))
+
+	depSet := make(map[Resource]struct{}, total)
```

Comment:
So this would now deduplicate dependencies repeatedly even when merging len(theirs) == 0  or len(ours) == 0. Given that these lists are short I think this is worth it; do some extra work but be more foolproof.